### PR TITLE
Update repo and example to React Native 0.73

### PR DESCRIPTION
**Pull request by Saadnajmi**
# Summary

Update the library and example app in this repo to React Native 0.73, which allows us to add a visionOS example app. As of https://github.com/react-native-datetimepicker/datetimepicker/pull/888 , dateTimePicker now has a minimum React Native version of 0.73. This PR will update the rest of the repo to reflect that. 

To do this, I did:

1. Use [@rnx-kit/align-deps](https://microsoft.github.io/rnx-kit/docs/tools/align-dep) to update the package to React Native 0.73 (minimum required for visionOS support)
2. Update the version (and template files) of `react-native-test-app` used to the latest major release
 3. As a consequence, E2E tests no longer work on Android new arch, so I disabled the CircleCI test for it. We already do this for iOS as well.

## Test Plan

visionOS test app boots:

Screenshot 2024-04-04 at 4 35 13 PM

iOS test app for me boots as well. I didn't test Android or Windows.

### What's required for testing (prerequisites)?

Xcode 15.2

### What are the steps to reproduce (after prerequisites)?

Just run `pod install` and run the visionOS test app as you would an iOS test app

## Compatibility

| OS      | Implemented |
| ------- | :---------: |
| iOS     |    ✅ ❌     |
| visionOS     |    ✅ ❌     |
| Android |    ✅ ❌     |

## Checklist

- [ ] I have tested this on a device and a simulator
- [ ] I added the documentation in `README.md`
- [ ] I updated the typed files (TS and Flow)
- [x] I added a sample use of the API in the example project (`example/App.js`)
- [ ] I have added automated tests, either in JS or e2e tests, as applicable

### CODE REVIEW

**Comment by Saadnajmi** [comment]
The second lock file (example/yarn.lock) can be removed if we update to yarn 4 and use yarn workspaces. Yarn 1 workspaces requires the root package to be private, which doesn't work for our case. Would there be interest in upgrading? I can make a PR if so.

**Comment by vonovak** [comment]
hello @Saadnajmi and thank you for the PR!
I'll be happy to merge this (though this PR touches some of the same files as https://github.com/react-native-datetimepicker/datetimepicker/pull/878 - and right now I'm not sure which PR will be ready and merged first.)

I just want to make sure of one thing - I want to be able to run the example app, and have hot reloading working when I make edits in the `src` folder. Right now I'm not sure if metro will pick up those changes, can you pls double check that?

Thank you

**Comment by Saadnajmi** [comment]
> hello @Saadnajmi and thank you for the PR! I'll be happy to merge this (though this PR touches some of the same files as #878 - and right now I'm not sure which PR will be ready and merged first.)
> 
> I just want to make sure of one thing - I want to be able to run the example app, and have hot reloading working when I make edits in the `src` folder. Right now I'm not sure if metro will pick up those changes, can you pls double check that?
> 
> Thank you

Hey, sure I'll check that out. It seems I have a couple of things to check out, as the CI is failing. 
A couple of questions:
1. This PR is large mostly because it updates the test app and moves the dependencies around to accommodate. However, to add visionOS support, we actually _only_ need to change the one line in the podspec (No changes to native or JS source were needed). If you prefer, I could just split that update out to its own PR? I normally try to pair updates with a test app change as well, but wanted to double check.
2. Are we OK with the "extra" updates of adding RNX-Kit configs (to make RN update easier) and potentially using Yarn 4? I can definitely remove those changes, though keeping with yarn 1 does mean I'll either need to edit the CI pipelines or the test app more to keep everything building.

**Comment by vonovak** [comment]
Hello,
I'm fine with having the changes in one PR.
Using yarn 4 is also good.

With rnx kit, can you point me to some docs around what specifically the things you added (metro config it seems?) do? I haven't used that before so unless there's some really good benefits I would probably prefer to not have so that me or anyone else doesn't need to understand one more thing to contribute. TY.

Luckily the CI is fairly solid in this package so I trust that when it's green everything works.

**Comment by vonovak** [comment]
@Saadnajmi this will be a shot in the dark, but would perhaps MSFT be interested in sponsoring the maintenance of the module?
I have to say that I don't use windows so I couldn't be too active on the development side, but for things like keeping RN up to date, CI green and doing the necessary changes that always show up from time to time (now bridgeless mode), I think I could save you time there 🙂.

**Comment by Saadnajmi** [comment]
@vonovak Sorry, I asked around a bit, and I don't think we're at a point where we can sponsor repos at this time:/ 

Meanwhile, I'm happy to spend time submitting PRs to fix issues / do some maintenance as they come up for stuff we care about (currently, that's adding visionOS support to several repos). Speaking of, I do hope the extra changes help with the maintenance, since adding visionOS support was really just a one line change but I figured it's worth trying to make the world better while we're at it 🙂 . For the windows development / maintenance, feel free to ping me or other Microsoft people and I can redirect to someone who does RN Windows dev (not me because I only really do Apple 😀 ). 

For this PR, I need to keep working on making the CI green. You can probably ignore till that's done and/or I ping you about it, does that work?

**Comment by vonovak** [comment]
Hey, thank you for the comment, works for me! 🙂 👍

**Comment by vonovak** [comment]
Yes, feel free to split the PR up 🙂.

As for codegen-generated code, if you find that the currently generated code is the same as the one we have checked-in in the repo, then I guess there's no need to have it checked in. But I find it unlikely. There were some customizations made on it, I don't remember exactly what it was, maybe some intricacies related to layout or state as you're saying.

**Comment by Saadnajmi** [comment]
> Yes, feel free to split the PR up 🙂.
> 
> As for codegen-generated code, if you find that the currently generated code is the same as the one we have checked-in in the repo, then I guess there's no need to have it checked in. But I find it unlikely. There were some customizations made on it, I don't remember exactly what it was, maybe some intricacies related to layout or state as you're saying.

@shwanton and I chatted offline about it. The repo adds custom state to the shadow note to track the date pickers' frame. That can't be replicated with _just_ codegen atm afaik, so it needs to stay checked in. I may be wrong

**Comment by vonovak** [comment]
Just got this message from circle:

In the last 30 days your team used the Free plan to test, build, and deploy for approximately 8752 minutes a day. That's a lot of building.

Without additional credits, your team will not be able to continue until the next refill on 05/22/24.

--

So in case building doesn't eun this might be the reason.

**Comment by Saadnajmi** [comment]
> Just got this message from circle:
> 
> In the last 30 days your team used the Free plan to test, build, and deploy for approximately 8752 minutes a day. That's a lot of building.
> 
> Without additional credits, your team will not be able to continue until the next refill on 05/22/24.
> 
> --
> 
> So in case building doesn't eun this might be the reason.

Ah, seems I have been going a bit too hard 😅 I can batch my updates better then. That does mean I probably won't revisit till 5/22/24.

File: package.json [hunk 1]
```diff
@@ -67,29 +67,31 @@
...
+    "flow-typed": "^3.9.0",
     "jest": "^29.5.0",
     "moment": "^2.24.0",
     "moment-timezone": "^0.5.41",
     "patch-package": "^6.4.7",
     "postinstall-postinstall": "^2.1.0",
     "prettier": "^2.8.8",
     "react": "18.2.0",
-    "react-native": "^0.72.7",
+    "react-native": "^0.73.0",
```
Comment (vonovak) [inline package.json:91]: Could we update to the latest RN 74 RC, so that bridgeless mode can also be tested?
Thank you

Comment (Saadnajmi) [reply package.json:91]: Could we not test bridgeless with 0.73? I think I'd be willing to do that after I can fix the CI for the current changes 😅 (Right now just New arch + Android fails)

Comment (vonovak) [reply package.json:89]: okay, let's do it separately. Keep this with 0.73 and I'll handle bridgeless with 0.74 separately 👍

File: package.json [hunk 2]
```diff
@@ -102,5 +104,13 @@
       "javaPackageName": "com.reactcommunity.rndatetimepicker"
     }
   },
-  "packageManager": "yarn@4.1.1"
+  "peerDependencies": {
+    "react": "18.2.0",
```
Comment (vonovak) [inline package.json:108]:
```suggestion
    "react": "*",
```

let's relax this a bit

File: package.json [hunk 3]
```diff
@@ -102,5 +104,13 @@
...
+    "react-native": "^0.73.0",
+    "react-native-windows": "^0.73.0"
+  },
+  "packageManager": "yarn@4.1.1",
+  "workspaces": [
+    "example"
+  ]
```
Comment (vonovak) [inline package.json:115]:
```suggestion
```

we're not using workspaces, are we?

File: package.json [hunk 4]
```diff
@@ -102,5 +104,13 @@
       "javaPackageName": "com.reactcommunity.rndatetimepicker"
     }
   },
-  "packageManager": "yarn@4.1.1"
+  "peerDependencies": {
+    "react": "18.2.0",
+    "react-native": "^0.73.0",
+    "react-native-windows": "^0.73.0"
+  },
+  "packageManager": "yarn@4.1.1",
```
Comment (vonovak) [inline package.json:112]:
```suggestion
  "peerDependenciesMeta": {
    "react-native-windows": {
      "optional": true
    }
  },
  "packageManager": "yarn@4.1.1",
```

**Review by vonovak — CHANGES_REQUESTED** [review]

**Review by vonovak — APPROVED** [review]
🚀 🚀 🚀 🚀 🚀 🚀 🚀 🚀

**Comment by vonovak** [comment]
:tada: This issue has been resolved in version 8.0.0 :tada:

If this package helps you, consider [sponsoring us](https://github.com/sponsors/react-native-datetimepicker)! :rocket:

**Comment by Saadnajmi** [comment]
> 🎉 This issue has been resolved in version 8.0.0 🎉
> 
> If this package helps you, consider [sponsoring us](https://github.com/sponsors/react-native-datetimepicker)! 🚀

Updating our rnx-kit preset :)
https://github.com/microsoft/rnx-kit/pull/3112

**Pull request merged**